Unix socket support implementation

sopvop opened the pull request:
This is not yet ready to be merged.
Tests and documentation are missing. But I think this PR can be used
as a starting point for implementation and discussion.

To use unix sockets specify "unix:/abs/path.sock" as address in config.
It is probably better to use separate parameter unix socket
specification. Server then can listen on http, https and unix sockets at
the same time.

Setting access permissions on socket is important if your snap server
and proxy, like nginx, are run with different users. So you can set
group write permission and add nginx user to same group as snap server.

On command line access rights are specified in octal.

Permissions are set with umask, so there is a race condition because
umask is process wide shared attribute. Setting permissions using
setFilePermissions can lead to race conditions if socket file is renamed
or unlinked while program is running.
In theory on Linux at least you can set permissions for socket Fd before
binding, but in my case it was setting not all permissions. Possibly
because it is also affected by umask set somewhere else.

Checking for HAS_SENDFILE, because it is defined on *nixes.

Using abs path only is a security feature, and is often recommended, for
example in TLPI.

gregorycollins commented:
First off, before reading any of the code:

> Checking for HAS_SENDFILE, because it is defined on *nixes.

I would add a new conditional HAS_UNIX_SOCKETS here.

gregorycollins commented on src/Snap/Http/Server.hs:
```diff
@@ -221,6 +221,7 @@ listeners conf = TLS.withTLS $ do
     httpListener = do
         p <- getPort conf
         b <- getBind conf
+        when ("unix:/" `S.isPrefixOf` b) Nothing
```
I think it's clearer as

```haskell
guard $ not $ "unix:/" `S.isPrefixOf` b
```

gregorycollins commented on src/Snap/Http/Server.hs:
```diff
@@ -229,6 +230,15 @@ listeners conf = TLS.withTLS $ do
+        b <- getBind conf
+        let acc = getUnixSocketAccessMode conf
+        if "unix:/" `S.isPrefixOf` b
+           then let path = S.unpack $ S.drop 5 b
```
Can't just unpack here, we need to take a round-trip through utf-8:

```haskell
unixListener = do
    b <- getBind conf
    let acc = getUnixSocketAccessMode conf
    guard $ "unix:/" `S.isPrefixOf` b
    let path = T.unpack $ T.decodeUtf8 $ S.drop 5 b
    return $! (b, do sock <- Sock.bindUnixSocket acc path
                  return $! (sock, Sock.httpAcceptFunc sock))
```

gregorycollins commented on src/Snap/Internal/Http/Server/Address.hs:
```diff
@@ -55,14 +55,14 @@ getAddress = getAddressImpl getHostAddr
+    SockAddrInet p _ -> host (fromIntegral p)
+    SockAddrInet6 p _ _ _ -> host (fromIntegral p)
+    SockAddrUnix path -> return (-1, S.pack $ "unix:" ++ path)
+    --x -> throwIO $ AddressNotSupportedException $ show x
```
Remove comment

gregorycollins commented on src/Snap/Internal/Http/Server/Config.hs:
```diff
@@ -236,6 +242,7 @@ instance Show (Config m a) where
                      , "sslcert: "        ++ _sslcert
                      , "sslchaincert: "   ++ _sslchaincert
                      , "sslkey: "         ++ _sslkey
+                     , "unixpemrs: "      ++ _unixacc
```
Typo

gregorycollins commented on src/Snap/Internal/Http/Server/Config.hs:
```diff
@@ -199,6 +204,7 @@ data Config m a = Config
     , sslcert        :: Maybe FilePath
     , sslchaincert   :: Maybe Bool
     , sslkey         :: Maybe FilePath
+    , unixacc        :: Maybe Int
```
Choose a more descriptive name?

sopvop commented:
What about adding another option to Config specifically for unix socket, instead of using bind?
This will make it possible to start server listening to https, http and unix socket. Like postgresql does.

gregorycollins commented on src/Snap/Internal/Http/Server/Socket.hs:
```diff
@@ -59,6 +63,29 @@ bindSocketImpl _setSocketOption _bindSocket _listen bindAddr bindPort = do
+      N.listen sock 150
+      return $! sock
+   where
+     isAbsPath = case path of
```
```haskell
System.FilePath.isAbsolute
```

sopvop replied:
that adds filepath dependency

gregorycollins commented on src/Snap/Internal/Http/Server/Socket.hs:
```diff
@@ -59,6 +63,29 @@ bindSocketImpl _setSocketOption _bindSocket _listen bindAddr bindPort = do
+   when (not isAbsPath) $
+      throwIO (AddressNotSupportedException path)
+   bracketOnError (socket N.AF_UNIX N.Stream 0) N.close $ \sock -> do
+      exists <- fileExist path
```
```haskell
eatException $ removeLink path
```
Don't bother with the stat() check.

sopvop replied:
This will also eat exception thrown if user has insufficient rights

gregorycollins replied:
You'll get it further below when you try to bind()?

gregorycollins commented on src/Snap/Internal/Http/Server/Socket.hs:
```diff
@@ -59,6 +63,29 @@ bindSocketImpl _setSocketOption _bindSocket _listen bindAddr bindPort = do
+         Nothing -> N.bindSocket sock (N.SockAddrUnix path)
+         Just mask -> bracket (setFileCreationMask mask)
+                              setFileCreationMask
+                              $ \_ -> N.bindSocket sock (N.SockAddrUnix path)
```
Prefer `const` to `\_ -> ...`.

gregorycollins commented:
Looks good to me so far!

> It is probably better to use separate parameter unix socket specification. Server then can listen on http, https and unix sockets at the same time.

Yep.

gregorycollins commented:
> What about adding another option to Config specifically for unix socket, instead of using bind?
This will make it possible to start server listening to https, http and unix socket. Like postgresql does.

Sounds good to me. However, reworking the configuration stuff is one of the final todos before we can release Snap 1.0.

At this point the internal interfaces are good in my opinion; what we give you is [rawHttpServe](https://github.com/snapframework/snap-server/blob/8bf6454b96035e7bfbb015ffc546a3d7fa2b6eff/src/Snap/Http/Server.hs#L69):

```haskell
rawHttpServe :: ServerHandler s  -- ^ server handler
             -> ServerConfig s   -- ^ server config
             -> [AcceptFunc]     -- ^ listening server backends
             -> IO ()
```

We need to expose more of this in a principled way and clean up the command line config stuff anyways.

sopvop commented:
Addressed issues pointed out by Gregory.
I decided to not add dependencies on filepath package for only isAbsolute.

Still needs tests and documentation.

sopvop commented:
Just realized that isAbsolute also checks for .. relative links in path.

gregorycollins commented:
> I decided to not add dependencies on filepath package for only isAbsolute.

We already depend on it transitively via snap-core.

sopvop commented:
Should I add anything else to this?

gregorycollins commented on src/Snap/Internal/Http/Server/Config.hs:
```diff
@@ -385,6 +404,24 @@ getSSLChainCert = sslchaincert
+--  to limit access to socket is creating it in a directory with proper permissions
+--  set.
+--
+--  Most BSD systems ignore access permissionse on sockets.
```
Typo here "permissionse"

gregorycollins commented on src/Snap/Internal/Http/Server/Config.hs:
```diff
@@ -385,6 +404,24 @@ getSSLChainCert = sslchaincert
+
+-- | Access mode for unix socket, by default is system specific.
+--  This should only be used to grant additional permissions to created
+--  socket file, and not to remove permissions set by default. The only portable way
```
Reflow the text in this paragraph?

gregorycollins commented on src/Snap/Internal/Http/Server/Socket.hs:
```diff
@@ -59,6 +68,25 @@ bindSocketImpl _setSocketOption _bindSocket _listen bindAddr bindPort = do
+#if HAS_UNIX_SOCKETS
+bindUnixSocket mode path = do
+   when (isRelative path) $
+      throwIO (AddressNotSupportedException path)
```
Give more info the user about why, e.g.:

```haskell
throwIO . AddressNotSupportedException $! "Refusing to bind unix socket to non-absolute path: " ++ path
```

gregorycollins commented:
Looks ready to merge from my perspective, except those few little quibbles I mentioned.

sopvop commented:
Is this ok?
